docs/dev/data_validation.rst

```diff
@@ -0,0 +1,694 @@
+completely null, since that is typically indicative of a bad ``ENUM`` constraint, a
+column naming error, or a bad merge, and should be investigated. To add this basic
+default, you add the test to the table level ``data_tests`` with no parameters, which
+building on the above example would look like:
+
+.. code-block:: yaml
+
+    version: 2
+    sources:
+      - name: pudl
+        tables:
+          - name: new_table_name
+            data_tests:
+              - expect_columns_not_all_null
+              - check_row_counts_per_partition:
+                  table_name: new_table_name
+                  partition_column: report_date
+
+--------------------------------------------------------------------------------
+Defining new data validation tests
+--------------------------------------------------------------------------------
+
+* How to define a new generic test (lean on references to dbt docs when possible)
+* Focus on the things that make the PUDL use case unusual.
+* DuckDB + Parquet means we can't rely on ``adapter`` object methods (no real DB)
+* Almost all our tables are "sources" not "models"
+
+Defining Macros
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* In dbt, macros are reusable SQL snippets that can be used to simplify your tests. You
+  can define a macro once and then use it in multiple tests. This is particularly useful
+  for complex tests that require a lot of boilerplate code.
+
+Testing the Tests
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* One reason to create macros for more complex functions is that they can be
+  independently unit-tested.
+
+Creating intermediate tables for a test
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. todo::
+
+  This section still seems a bit garbled. Clarify and confirm that the instructions are
+  complete and correct.
+
+In some cases you may need to modify a table or calculate some derived values before you
+can apply a test. This can be done by creating a new dbt `model
+<https://docs.getdbt.com/docs/build/models>`_ that materializes the intermediate table
+you want to execute tests on.  Defining a new model means adding a SQL file to
+``dbt/models/{data_source}/{table_name}/`` containing a ``SELECT`` statement. The
+results of that ``SELECT`` will constitute the new table.
+
+Now, add a SQL file to this directory named ``validate_{table_name}`` and define your
+model for producing the intermediate table here. Finally, add the model to the
+``schema.yml`` file and define tests exactly as you would for a ``source`` table. See
+``models/ferc1/out_ferc1__yearly_steam_plants_fuel_by_plant_sched402`` for an example of
+this pattern.
+
+Note: when adding a model, it will be stored as a SQL ``view`` in the file
+``$PUDL_OUTPUT/pudl_dbt_tests.duckdb``.
+
+--------------------------------------------------------------------------------
+Unmigrated Data Validation Docs (cannibalize)
+--------------------------------------------------------------------------------
```

Comment:
Drop this section now or later?

Reply:
It seemed like raw materials that someone might want to pull into some of the unfinished sections, or at least topics / cases to make sure the new version covered (or superceded) so I was leaving it? It's slightly unclean but this is the dev docs and (hopefully) we're going to finish the docs this/next week...